File: clients/rn/src/engine/item/ItemCodecs.ts
/**
 * Codecs for aspect and reagent item JSON stored in PhoneSoulItem.text.
 * TypeScript port of AspectItemCodec.java and ReagentItemCodec.java.
 */

import type { PhoneSoulItem } from './PhoneSoulItem';

// ---------------------------------------------------------------------------
// Aspect
// ---------------------------------------------------------------------------

export interface AspectDefinition {
  version: number;
  promptOverlay: string | null;
  vitalityShifts: Record<string, number>;
  selfDescription: string | null;
  slotHint: string;
  tokenEstimate: number;
}

/**
 * Decode a PhoneSoulItem's text field into an AspectDefinition.
 * Returns null if the item is not a valid aspect.
 */
export function decodeAspect(item: PhoneSoulItem): AspectDefinition | null {
  if (!item || !item.text || item.category !== 'aspect') return null;
  return decodeAspectJson(item.text);
}

/**
 * Decode a raw JSON string into an AspectDefinition.
 */
export function decodeAspectJson(json: string): AspectDefinition | null {
  if (!json) return null;
  try {
    const obj = JSON.parse(json);
    return {
      version: obj.version ?? 1,
      promptOverlay: obj.promptOverlay ?? null,
      vitalityShifts: obj.vitalityShifts ?? {},
      selfDescription: obj.selfDescription ?? null,
      slotHint: obj.slotHint ?? 'garment',
      tokenEstimate: obj.tokenEstimate > 0 ? obj.tokenEstimate : 20,
    };
  } catch {
    return null;
  }
}

/**
 * Encode an AspectDefinition to a JSON string (for PhoneSoulItem.text).
 */
export function encodeAspect(def: AspectDefinition): string {
  return JSON.stringify(def);
}

/**
 * Create a new AspectDefinition with defaults applied.
 */
export function createAspectDefinition(
  promptOverlay: string | null,
  vitalityShifts: Record<string, number>,
  selfDescription: string | null,
  slotHint: string,
  tokenEstimate: number,
): AspectDefinition {
  return {
    version: 1,
    promptOverlay,
    vitalityShifts: { ...vitalityShifts },
    selfDescription,
    slotHint: slotHint || 'garment',
    tokenEstimate: tokenEstimate > 0 ? tokenEstimate : 20,
  };
}

/** Whether the aspect injects prompt text. */
export function aspectHasPromptOverlay(def: AspectDefinition): boolean {
  return def.promptOverlay != null && def.promptOverlay.trim().length > 0;
}

/** Whether the aspect modifies vitality baselines. */
export function aspectHasVitalityShifts(def: AspectDefinition): boolean {
  return Object.keys(def.vitalityShifts).length > 0;
}

// ---------------------------------------------------------------------------
// Reagent
// ---------------------------------------------------------------------------

export interface ReagentDefinition {
  version: number;
  vitalityEffects: Record<string, number>;
  durationTicks: number;
  promptOverlay: string | null;
  consumable: boolean;
  tokenEstimate: number;
}

/** Maximum allowed duration: 1800 ticks (~30 minutes). */
export const REAGENT_MAX_DURATION = 1800;

/**
 * Decode a PhoneSoulItem's text field into a ReagentDefinition.
 * Returns null if the item is not a valid reagent.
 */
export function decodeReagent(item: PhoneSoulItem): ReagentDefinition | null {
  if (!item || !item.text || item.category !== 'reagent') return null;
  return decodeReagentJson(item.text);
}

/**
 * Decode a raw JSON string into a ReagentDefinition.
 */
export function decodeReagentJson(json: string): ReagentDefinition | null {
  if (!json) return null;
  try {
    const obj = JSON.parse(json);
    return {
      version: obj.version ?? 1,
      vitalityEffects: obj.vitalityEffects ?? {},
      durationTicks: obj.durationTicks > 0 ? obj.durationTicks : 300,
      promptOverlay: obj.promptOverlay ?? null,
      consumable: obj.consumable ?? true,
      tokenEstimate: obj.tokenEstimate > 0 ? obj.tokenEstimate : 10,
    };
  } catch {
    return null;
  }
}

/**
 * Encode a ReagentDefinition to a JSON string (for PhoneSoulItem.text).
 */
export function encodeReagent(def: ReagentDefinition): string {
  return JSON.stringify(def);
}

/**
 * Create a new ReagentDefinition with defaults applied.
 */
export function createReagentDefinition(
  vitalityEffects: Record<string, number>,
  durationTicks: number,
  promptOverlay: string | null,
  consumable: boolean,
  tokenEstimate: number,
): ReagentDefinition {
  return {
    version: 1,
    vitalityEffects: { ...vitalityEffects },
    durationTicks: Math.min(durationTicks > 0 ? durationTicks : 300, REAGENT_MAX_DURATION),
    promptOverlay,
    consumable,
    tokenEstimate: tokenEstimate > 0 ? tokenEstimate : 10,
  };
}

/** Whether the reagent injects prompt text while active. */
export function reagentHasPromptOverlay(def: ReagentDefinition): boolean {
  return def.promptOverlay != null && def.promptOverlay.trim().length > 0;
}

/** Clamped duration (respects MAX_DURATION). */
export function reagentEffectiveDuration(def: ReagentDefinition): number {
  return Math.min(def.durationTicks, REAGENT_MAX_DURATION);
}

File: clients/kmp/shared/src/commonMain/kotlin/org/wyrdsekai/app/engine/item/ItemCodecs.kt
package org.wyrdsekai.app.engine.item

import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

/**
 * Simplified SoulItem for the phone client.
 *
 * The server's SoulItem includes embedding, signature, and Instant timestamps.
 * PhoneSoulItem strips those to the minimum needed for equipment, prompt context,
 * and FamilyLocker sync (hash is the content-address).
 */
@Serializable
data class PhoneSoulItem(
    val hash: String,
    val category: String,
    val label: String,
    val text: String,
    val creatorDid: String,
    val created: Long, // epoch millis
    val significance: Double,
    val tags: List<String> = emptyList(),
) {
    companion object {
        /**
         * Compute a content-address hash for item text.
         * Uses Kotlin's hashCode for phone (lightweight, no JCE dependency in KMP common).
         * Format: "ph-" prefix + unsigned hex to distinguish from server SHA-256 hashes.
         */
        fun computeHash(text: String): String {
            // Use a simple but well-distributed hash for KMP common code.
            // Two rounds of FNV-1a-like mixing to reduce collisions vs raw hashCode.
            var h = 0x811c9dc5L
            for (ch in text) {
                h = h xor ch.code.toLong()
                h = (h * 0x01000193L) and 0xFFFFFFFFL
            }
            return "ph-${h.toString(16).padStart(8, '0')}"
        }

        /**
         * Create a PhoneSoulItem with auto-computed hash.
         */
        fun create(
            category: String,
            label: String,
            text: String,
            creatorDid: String,
            significance: Double,
            tags: List<String> = emptyList(),
        ): PhoneSoulItem = PhoneSoulItem(
            hash = computeHash(text),
            category = category,
            label = label,
            text = text,
            creatorDid = creatorDid,
            created = currentTimeMillis(),
            significance = significance,
            tags = tags,
        )
    }
}

// --- Aspect Codec ---

/**
 * Decoded aspect definition from a PhoneSoulItem's text field.
 * Mirrors server's AspectItemCodec.AspectDefinition.
 */
@Serializable
data class AspectDefinition(
    val version: Int = 1,
    val promptOverlay: String? = null,
    val vitalityShifts: Map<String, Double> = emptyMap(),
    val selfDescription: String? = null,
    val slotHint: String = "garment",
    val tokenEstimate: Int = 20,
) {
    /** Whether this aspect injects prompt text. */
    fun hasPromptOverlay(): Boolean =
        !promptOverlay.isNullOrBlank()

    /** Whether this aspect modifies vitality baselines. */
    fun hasVitalityShifts(): Boolean =
        vitalityShifts.isNotEmpty()
}

/**
 * Codec for aspect item JSON stored in PhoneSoulItem.text.
 */
object AspectItemCodec {

    private val json = Json {
        ignoreUnknownKeys = true
        encodeDefaults = true
    }

    /** Decode a PhoneSoulItem's text field into an AspectDefinition. */
    fun decode(item: PhoneSoulItem): AspectDefinition? {
        if (item.text.isBlank() || item.category != "aspect") return null
        return decode(item.text)
    }

    /** Decode a JSON string into an AspectDefinition. */
    fun decode(jsonString: String): AspectDefinition? {
        if (jsonString.isBlank()) return null
        return try {
            json.decodeFromString<AspectDefinition>(jsonString)
        } catch (_: Exception) {
            null
        }
    }

    /** Encode an AspectDefinition to JSON string (for PhoneSoulItem.text). */
    fun encode(def: AspectDefinition): String =
        json.encodeToString(def)

    /** Create a new AspectDefinition. */
    fun create(
        promptOverlay: String?,
        vitalityShifts: Map<String, Double>,
        selfDescription: String?,
        slotHint: String,
        tokenEstimate: Int,
    ): AspectDefinition = AspectDefinition(
        version = 1,
        promptOverlay = promptOverlay,
        vitalityShifts = vitalityShifts,
        selfDescription = selfDescription,
        slotHint = slotHint,
        tokenEstimate = tokenEstimate,
    )

    /** Build a PhoneSoulItem for a validated aspect. */
    fun toSoulItem(
        name: String,
        def: AspectDefinition,
        creatorDid: String,
        significance: Double,
    ): PhoneSoulItem {
        val text = encode(def)
        val tags = mutableListOf(name.lowercase().replace(' ', '-'))
        if (def.slotHint.isNotBlank()) tags.add(def.slotHint)
        if (!def.selfDescription.isNullOrBlank()) {
            for (word in def.selfDescription.lowercase().split("\\s+".toRegex())) {
                if (word.length > 3 && tags.size < 8) tags.add(word)
            }
        }
        return PhoneSoulItem.create(
            category = "aspect",
            label = name,
            text = text,
            creatorDid = creatorDid,
            significance = significance,
            tags = tags,
        )
    }
}

// --- Reagent Codec ---

/**
 * Decoded reagent definition from a PhoneSoulItem's text field.
 * Mirrors server's ReagentItemCodec.ReagentDefinition.
 */
@Serializable
data class ReagentDefinition(
    val version: Int = 1,
    val vitalityEffects: Map<String, Double> = emptyMap(),
    val durationTicks: Int = 300,
    val promptOverlay: String? = null,
    val consumable: Boolean = true,
    val tokenEstimate: Int = 10,
) {
    /** Whether this reagent injects prompt text while active. */
    fun hasPromptOverlay(): Boolean =
        !promptOverlay.isNullOrBlank()

    /** Clamped duration (respects MAX_DURATION). */
    fun effectiveDuration(): Int =
        minOf(durationTicks, MAX_DURATION)

    companion object {
        /** Maximum allowed duration: 1800 ticks (~30 minutes). */
        const val MAX_DURATION: Int = 1800
    }
}

/**
 * Codec for reagent item JSON stored in PhoneSoulItem.text.
 */
object ReagentItemCodec {

    private val json = Json {
        ignoreUnknownKeys = true
        encodeDefaults = true
    }

    /** Decode a PhoneSoulItem's text field into a ReagentDefinition. */
    fun decode(item: PhoneSoulItem): ReagentDefinition? {
        if (item.text.isBlank() || item.category != "reagent") return null
        return decode(item.text)
    }

    /** Decode a JSON string into a ReagentDefinition. */
    fun decode(jsonString: String): ReagentDefinition? {
        if (jsonString.isBlank()) return null
        return try {
            json.decodeFromString<ReagentDefinition>(jsonString)
        } catch (_: Exception) {
            null
        }
    }

    /** Encode a ReagentDefinition to JSON string (for PhoneSoulItem.text). */
    fun encode(def: ReagentDefinition): String =
        json.encodeToString(def)

    /** Create a new ReagentDefinition. */
    fun create(
        vitalityEffects: Map<String, Double>,
        durationTicks: Int,
        promptOverlay: String?,
        consumable: Boolean,
        tokenEstimate: Int,
    ): ReagentDefinition = ReagentDefinition(
        version = 1,
        vitalityEffects = vitalityEffects,
        durationTicks = minOf(durationTicks, ReagentDefinition.MAX_DURATION),
        promptOverlay = promptOverlay,
        consumable = consumable,
        tokenEstimate = tokenEstimate,
    )

    /** Build a PhoneSoulItem for a reagent. */
    fun toSoulItem(
        name: String,
        def: ReagentDefinition,
        creatorDid: String,
        significance: Double,
    ): PhoneSoulItem {
        val text = encode(def)
        val tags = mutableListOf(
            name.lowercase().replace(' ', '-'),
            "reagent",
        )
        if (def.consumable) tags.add("consumable")
        return PhoneSoulItem.create(
            category = "reagent",
            label = name,
            text = text,
            creatorDid = creatorDid,
            significance = significance,
            tags = tags,
        )
    }
}

// --- Time utility ---

internal fun currentTimeMillis(): Long = kotlin.time.Clock.System.now().toEpochMilliseconds()
